Add feature of holes to the Maze-activity in harder maps

nswarup14 opened the pull request.

nswarup14 committed:
Feature - add holes

Add a "Make risk" button that enables holes in the maze.

- random placement, without consideration of on or off path,

- displayed as red circles,

- when a player hits a hole, the hole is removed, and the player is sent
  back to start,

- number of holes increases with maze complexity,

Also collaborates with previous release of Maze, and can open journal
objects saved by previous release.

Co-authored-by: Emil Dudev <emildudev@gmail.com> (GCI 2013)
Co-authored-by: Swarup N <nswarup14@gmail.com> (GSoC 2019)
Co-authored-by: James Cameron <quozl@laptop.org> (maintainer)

rhl-bthr commented:
Code not reviewed, linked to #27. Next time, please force push.

You did not give a description or steps to test, so I made assumptions.
Why does the first level have no holes ?

nswarup14 commented:
@pro-panda 
I have not completely finished the feature. There are still some issues that need fixing. Once I complete it, I shall comment and report the user behaviour and roadblocks if any.

The reason first level has no holes is that, it is a very simple and a small map. Holes are implemented only in harder maps.

Thanks!

quozl commented:
I agree that reopening https://github.com/sugarlabs/maze-activity/pull/27 and pushing to the associated branch would be more appropriate.

Also, in future when creating a pull request that is not ready for review, please choose "draft".  There is apparently no method to degrade a ready-for-review pull request to draft.

Also, please write a test case, as requested in https://github.com/sugarlabs/maze-activity/issues/16.

nswarup14 commented:
Fixes #16 

**Changes/Additions**
1) `.gitignore`
2) `game.py`
3) `activity.py`
4) `maze.py`
5) `player.py`

**Implements**
1) Added `hole` toggle button next to `harder` and `easier` buttons for enabling holes. Holes are only in harder maps.
2) `hole` button settings are saved in the journal object
3) `trail button` settings are saved in the journal object
4) On sharing a maze with holes enabled, the holes are communicated to all players. When one player passes over a hole, he falls (the animation is the player face becoming smaller and smaller as he falls) and he begins the map from start. The hole is then replaced with a green tile, which serves as an indication to other players that there was a hole there. The hole disappears and the that information is communicated to other players, basically now the players can pass over that tile. 
I feel this is better than just removing the hole because, that would defeat the purpose of adding holes in the first place. A user would have no idea where the hole was before and thus he no longer has an opportunity to evade them, which was the motive behind the feature.
At the moment, the tile is green, but that can be changed.
5) I could not show the trails of all players in a maze, because the trail overlaps and becomes irrelevant.
6) I have updated the `icon` used for the `hole` button.
7) Also collaboration worked :smile: correctly this time. 

Tested on Sugar v0.114 live build on a VM.

@quozl  @pro-panda 
I'm sorry for not pushing to the same PR as before. I believed having a new PR might help track the issue better.
Yes, next time I shall choose the option when opening a PR just for adding commits and not immediate merging.

Updated checklist from #27 
- [x] selectable using a toggle button in the toolbar, next to the easier and harder buttons,
- [x] saved in the journal object, and;
- [x] communicated over the collaboration,
- [x] collaboration doesn't work quite right; the joined peer instance doesn't show the holes, but the 
original instance does; when the joined player passes over the hole position the two instances become out of sync with respect to player position,
- [x] the hole button is black on dark grey background of toolbar, and looks like it is missing; please use a white ring with black ring inside it, with transparent circle in centre,
- [x] draw the same hole icon in the maze instead of red squares,
- [x] your .giitignore file lacks a line ending,
- [x] use hole instead of blackhole for file name; risk of offending a minority, per code of conduct,
- [x] The mazes are still solvable.

Thanks!

quozl commented:
Thanks.

I'm puzzled by your thinking that a new pull request might help track the issue better.  Can you explain further what you mean?  Try to imagine my perspective.  For me as [a reviewer](https://github.com/sugarlabs/sugar-docs/blob/master/src/contributing.md#guide-for-reviewers) and a mentor, the problems with a new pull request are;
* when tracking your work by git hash, having to change my `git fetch origin pull/30/head` slightly,
* when tracking your work by branch, having to change my local clone branch again, instead of just `git pull`,
* having to look at more than one pull request at a time; the old one and the new one, or risk losing track of what was requested or reviewed in an old pull request,
* more noise added to an issue (each new pull request is mentioned in the issue),
* more noise in general for all Sugar Labs contributors.

Hope that helps explain the impact of new pull requests.  :grinning:

quozl commented on .gitignore:
```diff
@@ -1,2 +1,4 @@
 dist
 *.pyc
+settings.json
```
Why add settings.json to .gitignore?

quozl commented on activity.py:
```diff
@@ -107,6 +109,12 @@ def build_toolbar(self):
         toolbar_box.toolbar.insert(harder_button, -1)
 
+        self.hole_button = ToggleToolButton('maze-hole')
+        self.hole_button.set_tooltip(_('Add holes'))
```
Here you set the tooltip to "Add holes", but the function argument is "add_hole".  Plural or singular should be maintained if possible.

nswarup14 replied:
Yes, I have changed that.

quozl commented on activity.py:
```diff
@@ -133,6 +141,13 @@ def build_toolbar(self):
 
         return toolbar_box
 
+    def _add_hole(self, button):
```
Here self.game.hole could be assigned directly from button.get_active().  Once assigned, the next call to get_active() can be avoided.

nswarup14 replied:
i have refactored the code accordingly

quozl commented on activity.py:
```diff
@@ -227,7 +242,9 @@ def write_file(self, file_path):
                 'height': self.game.maze.height,
-                'finish_time': self.game.finish_time}
+                'finish_time': self.game.finish_time,
+                'hole': self.game.hole}
```
Three different terms for the same concept; self.game.hole is not "add_hole" or "add holes".

nswarup14 replied:
I have corrected this.

quozl commented on game.py:
```diff
@@ -545,7 +588,7 @@ def handleMessage(self, player, message):
                 Request to please send me the maze.  Reply is maze:.
 
-            maze: running_time, seed, width, height
+            maze: running_time, seed, width, height, hole
```
"hole" is not defined here.

nswarup14 replied:
@quozl 
I have corrected it and changed the variable name to `add_hole`.

quozl commented on game.py:
```diff
@@ -559,6 +602,13 @@ def handleMessage(self, player, message):
 
             show_trail: True/False
 
+            maze_hole: True/False
```
I'm counting.  Four different names for holes enabled.  :grin:

nswarup14 replied:
😄 Sorry about that. I defined the names as when I needed them and thus ended up with different names.
I have changed most of the variable names to `add_hole` for simplicity.

quozl commented on game.py:
```diff
@@ -590,8 +640,10 @@ def handleMessage(self, player, message):
+            running_time, seed, width, height, hole = map(lambda x: int(x),
+                                                          message[5:]
+                                                          .split(","))
+            self.hole = True if hole else False
```
This line seems very complicated.  Should this be `self.hole = bool(hole)` ?

nswarup14 replied:
Yes, I have corrected that.

quozl commented:
Thanks.  Reviewed.

* [ ] please properly credit Emil Dudev for any code or design concepts that you've reused; see [merge request](http://git.sugarlabs.org/maze/mainline/merge_requests/2)

quozl commented:
Tested 442d8e1, can't play a second maze, traceback;

```log
Traceback (most recent call last):
  File "/usr/share/sugar/activities/Maze.activity/game.py", line 698, in harder
    self._restart(newWidth, newHeight)
  File "/usr/share/sugar/activities/Maze.activity/game.py", line 712, in _restart
    add_hole=self.hole)
AttributeError: 'MazeGame' object has no attribute 'hole'
```

Collaboration worked fine.

nswarup14 commented:
@quozl 
I have fixed the traceback. The activity runs on restart, or choosing a harder maze
Thanks

quozl commented:
Thanks.

Reviewed 164e5534.  Please find my review as commits on [my branch _develop_](https://github.com/quozl/maze-activity/commits/develop).  Please review, and cherry-pick them in order, or pull from my branch.

I'll continue the review tomorrow with the intent of merging.

nswarup14 commented:
@quozl 
Thanks!
I agree to all the commits that you have added and I pulled you branch in my `develop`.

Please review.

quozl commented:
Please review and pull again.

https://github.com/quozl/maze-activity/commit/8c6466647f2e93bb58ba1891dc8e19c7a260a222

I'm continuing to work toward minimising the change against _master_ and fix problems I've seen.  Other than testing and reviewing, you can leave me to it for a few days.  :grin:  Maze is one of my critical activities.

nswarup14 commented:
@quozl 
Thanks. Reviewed till https://github.com/quozl/maze-activity/commit/8c6466647f2e93bb58ba1891dc8e19c7a260a222.

quozl commented:
Thanks, please pull my branch and push yours.

quozl commented:
Okay, I think I'm done.  Please review.

https://github.com/quozl/maze-activity/commit/d834f76a786719567f073cd6cf55e0dcaa2b12e0

My next step will be to collapse all commits into one, with appropriate Co-authored-by: lines for Emil, yourself, and myself.  Comments?

nswarup14 commented:
@quozl 
Reviewed till https://github.com/quozl/maze-activity/commit/d834f76a786719567f073cd6cf55e0dcaa2b12e0 and it looks good.

> My next step will be to collapse all commits into one, with appropriate Co-authored-by: lines for Emil, yourself, and myself. Comments?

I agree. No comments.

quozl commented:
Late change, supporting users who have not upgraded.  c9b219c

quozl commented:
Squashed and rewrote commit message.

@nswarup14, @pro-panda, @chimosky, @Aniket21mathur, please review or test.

Test case one;
* start Maze, observe new risk toggle button, click easier, harder, and risk button; maze shall regenerate each time, holes will increase with maze complexity,
* play maze; player shall fall into hole and restart, next passage over hole shall be harmless.

Test case two;
* start Maze, share, join,
* maze changes will be shared; easier, harder, and risk,
* play maze; only one player shall fall into any given hole.

nswarup14 commented:
Reviewed and tested https://github.com/sugarlabs/maze-activity/pull/30/commits/7f34b7b7f55c1178b4267c5efb33aa6c40705a3c with two VM instances running Sugar Live build with v0.114.

Test Case one:
Worked as expected. Clicking the `risk-toggle` button recreates a new maze with the holes added.

Test Case two:
Works as expected.

quozl committed:
Minimise redraws

With GTK_DEBUG=updates the maze was redrawn entirely each time a player
moved.  This slowed the activity and cost more battery power.

Port from Pygame may be a cause; the draw signal handler for
Gtk.DrawingArea MazeGame was overloaded with actions that were not
required for drawing the widget.

- use the draw signal handler only for painting the changed area from
  the cached surface, add a redraw function for redrawing the cached
  surface, and ensure that every maze change was marked as dirty,

After this was done, players falling into holes did so properly on the
local instance, but were animated in the wrong position on remote
instances.

Before risk holes were added, each instance was responsible for
animating a player move in the maze.  After risk holes were added, the
moving player instance was responsible for notifying others of falling
into a hole.  The notification did arrive at other instances before the
animation had finished moving to the hole coordinates.

- share the responsibility for detecting a fall, so that animation is
  exactly the same on each instance,

- remove the falling notification message; it is redundant.

quozl committed:
Share passed holes

When a maze with some passed holes is received by a newly joining
instance, that user is not shown the holes that have been passed.  When
they fall into the hole, they see themselves back at the start, but
others see them beyond the hole.

Share the list of passed holes when the maze definition is sent.

quozl commented:
Tested again and found two problems; redraws were slow, and joining a maze with some passed holes didn't show the holes as passed.  See 2b25718 and 9392efb commit messages for the technical detail.

chimosky commented:
Tested.

From the owner, the joined instance passes the hole and disappears but on the shared instance nothing happens and from the owner the other player is still seen even after the player passes the hole and disappear, also the holes don't show on the shared instance.

quozl commented:
@chimosky, thanks.  That doesn't happen for me.  It sounds like either inconsistent source code or network delays with collaboration.  Please check you are using the source code as at 9392efb on both instances.  Please debug further; apply this change to `activity.py`, ensure both systems show the same time of day, reproduce the problem, immediately stop the activity, and send us the logs;

```diff
--- a/activity.py
+++ b/activity.py
@@ -190,6 +190,7 @@ class MazeActivity(activity.Activity):
     def _received_cb(self, buddy, text):
         if buddy == self.owner:
             return
+        logging.error('rx: %s' % text.split('|')[1])
         self.game.msg_received(buddy, text)
 
     def _add_alert(self, title, text=None):
@@ -238,6 +239,7 @@ class MazeActivity(activity.Activity):
 
     def broadcast_msg(self, message):
         if self.text_channel:
+            logging.error('tx: %s' % message)
             # FIXME: can't identify the sender at the other end,
             # add the pubkey to the text message
             self.text_channel.send('%s|%s' % (self.my_key, message))
```

Here's what a sample run with a hole looks like; first from the logs of the owner or shared instance;

```log
1563528734.827195 ERROR root: rx: req_maze
1563528734.827499 ERROR root: tx: maze:24474256,1563528711,11,9,1
1563528734.828939 ERROR root: tx: move:1,1,0,0
1563528746.321003 ERROR root: rx: move:1,1,0,1
1563528748.550476 ERROR root: rx: move:5,3,1,0
1563528751.097011 ERROR root: tx: move:1,1,0,1
1563528753.084548 ERROR root: tx: move:5,3,1,0
1563528754.412781 ERROR root: tx: move:9,5,0,1
1563528754.518818 ERROR root: tx: finish:40.06
1563528756.237955 ERROR root: rx: move:1,1,0,1
1563528757.937822 ERROR root: rx: move:5,3,1,0
1563528758.848912 ERROR root: rx: move:9,5,0,1
1563528758.948088 ERROR root: rx: finish:24.08
Exited with status 0 [...]
```

Then from the logs of the joined instance;

```log
1563528734.796858 ERROR root: tx: req_maze
1563528734.846124 ERROR root: rx: maze:24474256,1563528711,11,9,1
1563528734.859871 ERROR root: rx: move:1,1,0,0
1563528746.296704 ERROR root: tx: move:1,1,0,1
1563528748.528140 ERROR root: tx: move:5,3,1,0
1563528751.103362 ERROR root: rx: move:1,1,0,1
1563528753.097211 ERROR root: rx: move:5,3,1,0
1563528754.427530 ERROR root: rx: move:9,5,0,1
1563528754.533173 ERROR root: rx: finish:40.06
1563528756.208106 ERROR root: tx: move:1,1,0,1
1563528757.919521 ERROR root: tx: move:5,3,1,0
1563528758.823748 ERROR root: tx: move:9,5,0,1
1563528758.929221 ERROR root: tx: finish:24.08
Exited with status 0 [...]
```

Also describe the environment you are using for testing; which virtualisation software if any, which Linux distribution release, and what network connection between the Sugar instances.

nswarup14 commented:
@chimosky
Doesn't happen with me either and activity runs as normal.

chimosky commented:
Tested again, works as expected. 

> Please check you are using the source code as at 9392efb on both instances.

Both instances were at 9392efb.

quozl commented:
I don't like inexplicable inconsistent results.  Let's wait for more testing.
